Cython/Compiler/Symtab.py

```diff
@@ -2254,6 +2259,11 @@ def __init__(self, name, outer_scope):
         self.class_name = name
         self.doc = None
 
+    def declare_builtin(self, name, pos):
+        if name in self.special_class_names:
+            return self.declare_var(name, py_object_type, pos)
+        return super(ClassScope, self).declare_builtin(name, pos)
+
```

Comment:
I don't get it. Why are we trying to declare builtins here? That seems wrong.

Reply:
I'm having another look at this because I mostly agree it doesn't seem quite right. But the logic is:

* when these were `scope_predefined_names` they were pre-declared in the scope. That meant that it wasn't possible to declare a different definition of them, because the initial declaration won.
* `declare_builtin` is called by `NameNode.analyse_types` as a last resort if no other definition is found. That works here because an overriding definition takes priority, but the names get found if no overriding definition is used.
* I don't feel too bad because these names are special magic names that are kind of close to builtins in some ways.

But I'll have more of a think about if there's a better way to do this.

-----------------------------------------------

I guess there's a secondary question about whether declaring a property to override `__module__` is even something we should allow.

The basic problem is that:

```
cdef class C:
     __module__ = "xxx"
     print(__module__)  # prints "xxx"
print(C.__module__)  # prints the original value
```

Declaring a property feels like it's just a workaround for that problem. And maybe that is the thing to solve instead.